# Wikidata にある日本の市区町村データの整合性チェック

Wikidata にある日本の市区町村データと、総務省サイトにあるそれとを比較し、一致することを確認する。

In [14]:
import pandas as pd
from SPARQLWrapper import SPARQLWrapper, JSON, XML

# 表のすべての行を表示する
pd.set_option('display.max_rows', None)

In [2]:
# 総務省のサイト https://www.soumu.go.jp/denshijiti/code.html から
# 最新の「都道府県コード及び市区町村コード」を読み込む。
excel_file_url = "https://www.soumu.go.jp/main_content/000894847.xls"

# １枚目のシート。市町村と東京特別区。普通の区は含まない。
df_municipalities_sub = pd.read_excel(excel_file_url, sheet_name=0)
# カラム名を変更
df_municipalities_sub.rename(columns={'都道府県名\n（漢字）': '都道府県名', '市区町村名\n（漢字）': '市区町村名'}, inplace=True)
# 市区町村名のカラムが　NaN の行は都道府県なので削除する
df_municipalities_sub.dropna(subset=['市区町村名'], inplace=True)
# 市
df_cities = df_municipalities_sub.query('市区町村名.str.endswith("市")')
# 東京特別区
df_tokyo_wards = df_municipalities_sub.query('市区町村名.str.endswith("区")')
# 町
df_towns = df_municipalities_sub.query('市区町村名.str.endswith("町")')
# 村
df_villages = df_municipalities_sub.query('市区町村名.str.endswith("村")')

# ２枚目のシート。普通の区。
df_wards = pd.read_excel(excel_file_url, sheet_name=1)
# カラム名を変更
df_wards.rename(columns={'都道府県名\n（漢字）': '都道府県名', '市区町村名\n（漢字）': '市区町村名', '都道府県名\n（ｶﾅ）': '都道府県名\n（カナ）', '市区町村名\n（ｶﾅ）': '市区町村名\n（カナ）'}, inplace=True)
# 市区町村名のカラムが「市」で終わる行は市なので削除する
df_wards = df_wards.query('not 市区町村名.str.endswith("市")')

print('市の数: ' + str(len(df_cities)))
print('東京特別区の数: ' + str(len(df_tokyo_wards)))
print('町の数: ' + str(len(df_towns)))
print('村の数: ' + str(len(df_villages)))
print('区の数: ' + str(len(df_wards)))

# 二つを合わせてすべての自治体のリストを得る
#df_municipalities = pd.concat([df_municipalities_sub, df_wards],ignore_index=True)
#df_municipalities.sort_values('団体コード', ascending=True, inplace=True)
#df_municipalities.reset_index(inplace=True, drop=True)
#df_municipalities

市の数: 792
東京特別区の数: 23
町の数: 743
村の数: 189
区の数: 171


市の数: 792
東京特別区の数: 23
町の数: 743
村の数: 189（北方領土の6村を含む）
区の数: 171（2024/1/1 の浜松市の区統合（7区→3区）後の数を先取りしている。現在の175-7+3=171）

が正しい。（2023/10/03 現在）

In [4]:
# Wikidata から SPARQL 言語による検索結果を RDF(JSON) で得る関数を定義する。
def QuerySparql(query):
    sparql = SPARQLWrapper("https://query.wikidata.org/sparql")
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    results = sparql.queryAndConvert()
    return results['results']['bindings']

## 都道府県ごとの市の数のチェック

In [30]:
# 都道府県ごとの市の数
num_cities_query = """
SELECT ?pref ?prefLabel ?prefCode (COUNT(DISTINCT ?city) AS ?count) WHERE {
  ?city wdt:P31/wdt:P279* wd:Q494721 . # ?city は「日本の市」またはその部分集合（の部分集合……）の要素である。
  MINUS { ?city wdt:P576 ?abolished. } # ただし「廃止日」がないこと。
  OPTIONAL {
    # P131 は「位置する行政区画」を表す。
    # 以下のような例で正しく都道府県を得るために、wdt:P131 のあとに '+' が必要。
    # 札幌市 P131 石狩振興局 P131 北海道
    ?city wdt:P131+ ?pref. # ?pref に位置する。
    ?pref wdt:P31 wd:Q50337. # ?pref は「都道府県」である。
  }
  OPTIONAL { ?pref wdt:P429 ?prefCode. } # 都道府県の全国地方公共団体コード
  SERVICE wikibase:label {
    bd:serviceParam wikibase:language "ja" .
  }
}
GROUP BY ?pref ?prefLabel ?prefCode
ORDER BY ASC(?prefCode)
"""
json_num_cities_wikidata = QuerySparql(num_cities_query)

# pandas DataFrame へ変換
num_cities_wikidata = []
for c in json_num_cities_wikidata:
    city = {
        '都道府県名': c['prefLabel']['value'],
        '市の数': int(c['count']['value'])
    }
    num_cities_wikidata.append(city)
df_num_cities_wikidata = pd.DataFrame(num_cities_wikidata)
df_num_cities_wikidata

都道府県名  市の数
0    北海道   35
1    青森県   10
2    岩手県   14
3    宮城県   14
4    秋田県   13
5    山形県   13
6    福島県   13
7    茨城県   32
8    栃木県   14
9    群馬県   12
10   埼玉県   40
11   千葉県   37
12   東京都   26
13  神奈川県   19
14   新潟県   20
15   富山県   10
16   石川県   11
17   福井県    9
18   山梨県   13
19   長野県   19
20   岐阜県   21
21   静岡県   23
22   愛知県   38
23   三重県   14
24   滋賀県   13
25   京都府   15
26   大阪府   33
27   兵庫県   29
28   奈良県   12
29  和歌山県    9
30   鳥取県    4
31   島根県    8
32   岡山県   15
33   広島県   14
34   山口県   13
35   徳島県    8
36   香川県    8
37   愛媛県   11
38   高知県   11
39   福岡県   29
40   佐賀県   10
41   長崎県   13
42   熊本県   14
43   大分県   14
44   宮崎県    9
45  鹿児島県   19
46   沖縄県   11

これらの Wikidata による市の数を、総務省のページ
https://www.j-lis.go.jp/spd/code-address/kenbetsu-inspection/cms_11914151.html
にある数と比べ、一致することを確認すること。

In [5]:
# 市の一覧を Wikidata から取得
cities_query = """
SELECT DISTINCT ?pref ?prefLabel ?city ?cityLabel ?code WHERE {
  {
    # 以下のような例を得るために、wdt:P31 のあとに "/wdt:P279*" が必要。
    # 山形市 P31 中核市 P279 日本の市
    # 所沢市 P31 特例市 P279 日本の市
    ?city wdt:P31/wdt:P279* wd:Q494721 . # ?city は「日本の市」の（直接または間接的な）要素である。
  } 
  MINUS { ?city wdt:P576 ?abolished . } # ただし「廃止日」がないこと。

  OPTIONAL { ?city wdt:P429 ?code . } # 全国地方公共団体コード
  OPTIONAL {
    # 以下のような例で正しく都道府県を得るために、wdt:P131 のあとに '+' が必要。
    # 札幌市 P131 石狩振興局 P131 北海道
    ?city wdt:P131+ ?pref. # ?pref に（直接または間接的に）位置する。
    ?pref wdt:P31 wd:Q50337. # ?pref は「都道府県」である。
    MINUS { ?pref wdt:P576 ?abolished . } # ただし「廃止日」がないこと。
  }

  SERVICE wikibase:label { bd:serviceParam wikibase:language "ja" . }
}
ORDER BY ASC(?code)
"""
json_cities_wikidata = QuerySparql(cities_query)

# pandas DataFrame へ変換
cities_wikidata = []
for item in json_cities_wikidata:
    city = {
        '団体コード': (int(item['code']['value']) if ('code' in item) else -1),
        '都道府県名': (item['prefLabel']['value'] if ('prefLabel' in item) else None),
        '市名': (item['cityLabel']['value'] if ('cityLabel' in item) else None)
    }
    cities_wikidata.append(city)
df_cities_wikidata = pd.DataFrame(cities_wikidata)
df_cities_wikidata

団体コード 都道府県名    市名
0     11002   北海道   札幌市
1     12025   北海道   函館市
2     12033   北海道   小樽市
3     12041   北海道   旭川市
4     12050   北海道   室蘭市
..      ...   ...   ...
787  472115   沖縄県   沖縄市
788  472123   沖縄県  豊見城市
789  472131   沖縄県  うるま市
790  472140   沖縄県  宮古島市
791  472158   沖縄県   南城市

[792 rows x 3 columns]

In [6]:
# Wikidata にあって総務省のデータにはないものを表示
# 何もないのが正しい。
df_cities_only_in_wikidata = df_cities_wikidata[~df_cities_wikidata['団体コード'].isin(df_cities['団体コード'])]
df_cities_only_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 市名]
Index: []

In [7]:
# 総務省のデータにあって Wikidata にはないものを表示
# 何もないのが正しい。
df_cities_not_in_wikidata = df_cities[~df_cities['団体コード'].isin(df_cities_wikidata['団体コード'])]
df_cities_not_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 市区町村名, 都道府県名
（カナ）, 市区町村名
（カナ）]
Index: []

In [9]:
# 東京特別区の一覧を Wikidata から取得
tokyo_wards_query = """
SELECT DISTINCT ?pref ?prefLabel ?ward ?wardLabel ?code WHERE {
  ?ward wdt:P31 wd:Q5327704 . # ?ward は「日本の特別区」の要素である。
  MINUS { ?ward wdt:P576 ?abolished . } # ただし「廃止日」がないこと。

  OPTIONAL { ?ward wdt:P429 ?code . } # 全国地方公共団体コード
  OPTIONAL {
    # 以下のような例で正しく都道府県を得るために、wdt:P131 のあとに '+' が必要。
    # 札幌市 P131 石狩振興局 P131 北海道
    ?ward wdt:P131+ ?pref. # ?pref に（直接または間接的に）位置する。
    ?pref wdt:P31 wd:Q50337. # ?pref は「都道府県」である。
    MINUS { ?pref wdt:P576 ?abolished . } # ただし「廃止日」がないこと。
  }

  SERVICE wikibase:label { bd:serviceParam wikibase:language "ja" . }
}
ORDER BY ASC(?code)
"""
json_tokyo_wards_wikidata = QuerySparql(tokyo_wards_query)

# pandas DataFrame へ変換
tokyo_wards_wikidata = []
for item in json_tokyo_wards_wikidata:
    ward = {
        '団体コード': (int(item['code']['value']) if ('code' in item) else -1),
        '都道府県名': (item['prefLabel']['value'] if ('prefLabel' in item) else None),
        '区名': (item['wardLabel']['value'] if ('wardLabel' in item) else None)
    }
    tokyo_wards_wikidata.append(ward)
df_tokyo_wards_wikidata = pd.DataFrame(tokyo_wards_wikidata)
df_tokyo_wards_wikidata

団体コード 都道府県名    区名
0   131016   東京都  千代田区
1   131024   東京都   中央区
2   131032   東京都    港区
3   131041   東京都   新宿区
4   131059   東京都   文京区
5   131067   東京都   台東区
6   131075   東京都   墨田区
7   131083   東京都   江東区
8   131091   東京都   品川区
9   131105   東京都   目黒区
10  131113   東京都   大田区
11  131121   東京都  世田谷区
12  131130   東京都   渋谷区
13  131148   東京都   中野区
14  131156   東京都   杉並区
15  131164   東京都   豊島区
16  131172   東京都    北区
17  131181   東京都   荒川区
18  131199   東京都   板橋区
19  131202   東京都   練馬区
20  131211   東京都   足立区
21  131229   東京都   葛飾区
22  131237   東京都  江戸川区

In [10]:
# Wikidata にあって総務省のデータにはないものを表示
# 何もないのが正しい。
df_tokyo_wards_only_in_wikidata = df_tokyo_wards_wikidata[~df_tokyo_wards_wikidata['団体コード'].isin(df_tokyo_wards['団体コード'])]
df_tokyo_wards_only_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 区名]
Index: []

In [11]:
# 総務省のデータにあって Wikidata にはないものを表示
# 何もないのが正しい。
df_tokyo_wards_not_in_wikidata = df_tokyo_wards[~df_tokyo_wards['団体コード'].isin(df_tokyo_wards_wikidata['団体コード'])]
df_tokyo_wards_not_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 市区町村名, 都道府県名
（カナ）, 市区町村名
（カナ）]
Index: []

In [15]:
# 区の一覧を Wikidata から取得
wards_query = """
SELECT DISTINCT ?pref ?prefLabel ?city ?cityLabel  ?ward ?wardLabel ?code WHERE {
  {?ward wdt:P31 wd:Q137773 . } # ?ward は「日本の区」の要素である。
  MINUS { ?ward wdt:P576 ?abolished . } # ただし「廃止日」がないこと。

  OPTIONAL { ?ward wdt:P429 ?code . } # 全国地方公共団体コード
  OPTIONAL {
    # P131 は「位置する行政区画」を表す。
    ?ward wdt:P131 ?city. # ?ward は ?city に位置する。
    ?city wdt:P31 wd:Q494721. # ?city は「日本の市」である。
    MINUS { ?city wdt:P576 ?abolished . } # ただし「廃止日」がないこと。
  }
  OPTIONAL {
    # 以下のような例で正しく都道府県を得るために、wdt:P131 のあとに '+' が必要。
    # 札幌市 P131 石狩振興局 P131 北海道
    ?ward wdt:P131+ ?pref. # ?pref に（直接または間接的に）位置する。
    ?pref wdt:P31 wd:Q50337. # ?pref は「都道府県」である。
    MINUS { ?pref wdt:P576 ?abolished . } # ただし「廃止日」がないこと。
  }

  SERVICE wikibase:label { bd:serviceParam wikibase:language "ja" . }
}
ORDER BY ASC(?code)
"""
json_wards_wikidata = QuerySparql(wards_query)

# pandas DataFrame へ変換
wards_wikidata = []
for item in json_wards_wikidata:
    ward = {
        '団体コード': (int(item['code']['value']) if ('code' in item) else -1),
        '都道府県名': (item['prefLabel']['value'] if ('prefLabel' in item) else None),
        '市名': (item['cityLabel']['value'] if ('cityLabel' in item) else None),
        '区名': (item['wardLabel']['value'] if ('wardLabel' in item) else None)
    }
    wards_wikidata.append(ward)
df_wards_wikidata = pd.DataFrame(wards_wikidata)
df_wards_wikidata

団体コード 都道府県名     市名          区名
0     11011   北海道    札幌市         中央区
1     11029   北海道    札幌市          北区
2     11037   北海道    札幌市          東区
3     11045   北海道    札幌市         白石区
4     11053   北海道    札幌市         豊平区
5     11061   北海道    札幌市    南区 (札幌市)
6     11070   北海道    札幌市    西区 (札幌市)
7     11088   北海道    札幌市         厚別区
8     11096   北海道    札幌市         手稲区
9     11100   北海道    札幌市         清田区
10    41017   宮城県    仙台市   青葉区 (仙台市)
11    41025   宮城県    仙台市        宮城野区
12    41033   宮城県    仙台市         若林区
13    41041   宮城県    仙台市         太白区
14    41050   宮城県    仙台市          泉区
15   111015   埼玉県  さいたま市  西区 (さいたま市)
16   111023   埼玉県  さいたま市          北区
17   111031   埼玉県  さいたま市         大宮区
18   111040   埼玉県  さいたま市         見沼区
19   111058   埼玉県  さいたま市         中央区
20   111066   埼玉県  さいたま市          桜区
21   111074   埼玉県  さいたま市         浦和区
22   111082   埼玉県  さいたま市          南区
23   111091   埼玉県  さいたま市          緑区
24   111104   埼玉県  さいたま市         岩槻区
25   121011   千葉県    千葉市         中央区
26   121029   千葉県    千葉市        花見川区
27   121037   千葉県    千葉市         稲毛区
28   121045   千葉県    千葉市         若葉区
29   121053   千葉県    千葉市          緑区
30   121061   千葉県    千葉市         美浜区
31   141011  神奈川県    横浜市   鶴見区 (横浜市)
32   141020  神奈川県    横浜市        神奈川区
33   141038  神奈川県    横浜市          西区
34   141046  神奈川県    横浜市          中区
35   141054  神奈川県    横浜市          南区
36   141062  神奈川県    横浜市       保土ケ谷区
37   141071  神奈川県    横浜市         磯子区
38   141089  神奈川県    横浜市         金沢区
39   141097  神奈川県    横浜市         港北区
40   141101  神奈川県    横浜市         戸塚区
41   141119  神奈川県    横浜市         港南区
42   141127  神奈川県    横浜市          旭区
43   141135  神奈川県    横浜市    緑区 (横浜市)
44   141143  神奈川県    横浜市         瀬谷区
45   141151  神奈川県    横浜市          栄区
46   141160  神奈川県    横浜市          泉区
47   141178  神奈川県    横浜市         青葉区
48   141186  神奈川県    横浜市         都筑区
49   141313  神奈川県    川崎市         川崎区
50   141321  神奈川県    川崎市          幸区
51   141330  神奈川県    川崎市         中原区
52   141348  神奈川県    川崎市         高津区
53   141356  神奈川県    川崎市         多摩区
54   141364  神奈川県    川崎市         宮前区
55   141372  神奈川県    川崎市         麻生区
56   141518  神奈川県   相模原市          緑区
57   141526  神奈川県   相模原市         中央区
58   141534  神奈川県   相模原市   南区 (相模原市)
59   151017   新潟県    新潟市          北区
60   151025   新潟県    新潟市          東区
61   151033   新潟県    新潟市   中央区 (新潟市)
62   151041   新潟県    新潟市         江南区
63   151050   新潟県    新潟市         秋葉区
64   151068   新潟県    新潟市          南区
65   151076   新潟県    新潟市    西区 (新潟市)
66   151084   新潟県    新潟市         西蒲区
67   221015   静岡県    静岡市          葵区
68   221023   静岡県    静岡市         駿河区
69   221031   静岡県    静岡市         清水区
70   221317   静岡県    浜松市          中区
71   221325   静岡県    浜松市          東区
72   221333   静岡県    浜松市          西区
73   221341   静岡県    浜松市          南区
74   221350   静岡県    浜松市          北区
75   221368   静岡県    浜松市         浜北区
76   221376   静岡県    浜松市         天竜区
77   221384   静岡県    浜松市         中央区
78   221392   静岡県    浜松市         浜名区
79   231011   愛知県   名古屋市         千種区
80   231029   愛知県   名古屋市          東区
81   231037   愛知県   名古屋市          北区
82   231045   愛知県   名古屋市          西区
83   231053   愛知県   名古屋市         中村区
84   231061   愛知県   名古屋市          中区
85   231070   愛知県   名古屋市         昭和区
86   231088   愛知県   名古屋市         瑞穂区
87   231096   愛知県   名古屋市         熱田区
88   231100   愛知県   名古屋市         中川区
89   231118   愛知県   名古屋市          港区
90   231126   愛知県   名古屋市          南区
91   231134   愛知県   名古屋市         守山区
92   231142   愛知県   名古屋市          緑区
93   231151   愛知県   名古屋市         名東区
94   231169   愛知県   名古屋市         天白区
95   261017   京都府    京都市          北区
96   261025   京都府    京都市         上京区
97   261033   京都府    京都市         左京区
98   261041   京都府    京都市         中京区
99   261050   京都府    京都市         東山区
100  261068   京都府    京都市         下京区
101  261076   京都府    京都市          南区
102  261084   京都府    京都市         右京区
103  261092   京都府    京都市         伏見区
104  261106   京都府    京都市         山科区
105  261114   京都府    京都市         西京区
106  271021   大阪府    大阪市         都島区
107  27103

浜松市の 2024/1/1 の区統合によってできる中央区と浜名区が含まれるので、
現在の実数 175 よりも 2 多い 177 行がヒットする。

In [16]:
# Wikidata にあって総務省のデータにはないものを表示
# 何もないのが正しい。
#
# 2023/10/01 現在、浜松市の2024年に予定されている統合前の区7つがここに表示される。
df_wards_only_in_wikidata = df_wards_wikidata[~df_wards_wikidata['団体コード'].isin(df_wards['団体コード'])]
df_wards_only_in_wikidata

団体コード 都道府県名   市名   区名
70  221317   静岡県  浜松市   中区
71  221325   静岡県  浜松市   東区
72  221333   静岡県  浜松市   西区
73  221341   静岡県  浜松市   南区
74  221350   静岡県  浜松市   北区
75  221368   静岡県  浜松市  浜北区
76  221376   静岡県  浜松市  天竜区

In [17]:
# 総務省のデータにあって Wikidata にはないものを表示
# 何もないのが正しい。
#
# 2023/10/01 現在、浜松市の2024年に予定されている統合で団体コードがつけ変わる天竜区がここに表示される。
df_wards_not_in_wikidata = df_wards[~df_wards['団体コード'].isin(df_wards_wikidata['団体コード'])]
df_wards_not_in_wikidata

団体コード 都道府県名   市区町村名 都道府県名\n（カナ）  市区町村名\n（カナ） Unnamed: 5
82  221406   静岡県  浜松市天竜区     ｼｽﾞｵｶｹﾝ  ﾊﾏﾏﾂｼﾃﾝﾘｭｳｸ   R6.1.1施行

In [19]:
# 町の一覧を Wikidata から取得
towns_query = """
SELECT DISTINCT ?pref ?prefLabel ?district ?districtLabel ?town ?townLabel ?code WHERE {
  ?town p:P31 ?statement2 .
  ?statement2 ps:P31 wd:Q1059478 . # ?town は「日本の町」である。
  MINUS { ?statement2 pq:P582 ?endTime. } # 「終了日」の修飾子がないこと。
  MINUS { ?town wdt:P576 ?abolished . } # 「廃止日」がないこと。

  OPTIONAL { ?town wdt:P429 ?code . } # 全国地方公共団体コード
  OPTIONAL {
    # 以下のような例で正しく都道府県を得るために、wdt:P131 のあとに '+' が必要。
    # 札幌市 P131 石狩振興局 P131 北海道
    ?town wdt:P131+ ?pref. # ?pref に（直接または間接的に）位置する。
    ?pref wdt:P31 wd:Q50337. # ?pref は「都道府県」である。
    MINUS { ?pref wdt:P576 ?abolished . } # ただし「廃止日」がないこと。
  }
  OPTIONAL {
    ?town p:P131 ?statement1.
    ?statement1 ps:P131 ?district. # ?town は ?district に位置する。
    ?district wdt:P31 wd:Q1122846. # ?district は「郡」である。
    MINUS { ?statement1 pq:P582 ?endTime. } # 「終了日」が修飾子にあるものは除く。
  }

  SERVICE wikibase:label { bd:serviceParam wikibase:language "ja" . }
}
ORDER BY ASC(?code)
"""
json_towns_wikidata = QuerySparql(towns_query)

# pandas DataFrame へ変換
towns_wikidata = []
for item in json_towns_wikidata:
    town = {
        '団体コード': (int(item['code']['value']) if ('code' in item) else -1),
        '都道府県名': (item['prefLabel']['value'] if ('prefLabel' in item) else None),
        '郡名': (item['districtLabel']['value'] if ('districtLabel' in item) else None),
        '町名': (item['townLabel']['value'] if ('townLabel' in item) else None)
    }
    towns_wikidata.append(town)
df_towns_wikidata = pd.DataFrame(towns_wikidata)
df_towns_wikidata

団体コード 都道府県名     郡名         町名
0     13030   北海道    石狩郡        当別町
1     13315   北海道    松前郡        松前町
2     13323   北海道    松前郡        福島町
3     13331   北海道    上磯郡        知内町
4     13340   北海道    上磯郡       木古内町
5     13374   北海道    亀田郡        七飯町
6     13439   北海道    茅部郡        鹿部町
7     13455   北海道    茅部郡         森町
8     13463   北海道    二海郡        八雲町
9     13471   北海道    山越郡       長万部町
10    13617   北海道    檜山郡        江差町
11    13625   北海道    檜山郡       上ノ国町
12    13633   北海道    檜山郡       厚沢部町
13    13641   北海道    爾志郡        乙部町
14    13676   北海道    奥尻郡        奥尻町
15    13706   北海道    瀬棚郡        今金町
16    13714   北海道    久遠郡       せたな町
17    13927   北海道    寿都郡        寿都町
18    13935   北海道    寿都郡       黒松内町
19    13943   北海道    磯谷郡        蘭越町
20    13951   北海道    虻田郡       ニセコ町
21    13986   北海道    虻田郡       喜茂別町
22    13994   北海道    虻田郡        京極町
23    14001   北海道    虻田郡       倶知安町
24    14010   北海道    岩内郡        共和町
25    14028   北海道    岩内郡        岩内町
26    14052   北海道    積丹郡        積丹町
27    14061   北海道    古平郡        古平町
28    14079   北海道    余市郡        仁木町
29    14087   北海道    余市郡        余市町
30    14231   北海道    空知郡        南幌町
31    14249   北海道    空知郡       奈井江町
32    14257   北海道    空知郡       上砂川町
33    14273   北海道    夕張郡        由仁町
34    14281   北海道    夕張郡        長沼町
35    14290   北海道    夕張郡        栗山町
36    14303   北海道    樺戸郡        月形町
37    14311   北海道    樺戸郡        浦臼町
38    14320   北海道    樺戸郡      新十津川町
39    14338   北海道    雨竜郡       妹背牛町
40    14346   北海道    雨竜郡       秩父別町
41    14362   北海道    雨竜郡        雨竜町
42    14371   北海道    雨竜郡        北竜町
43    14389   北海道    雨竜郡        沼田町
44    14524   北海道    上川郡        鷹栖町
45    14532   北海道    上川郡       東神楽町
46    14541   北海道    上川郡        当麻町
47    14559   北海道    上川郡        比布町
48    14567   北海道    上川郡        愛別町
49    14575   北海道    上川郡        上川町
50    14583   北海道    上川郡        東川町
51    14591   北海道    上川郡        美瑛町
52    14605   北海道    空知郡      上富良野町
53    14613   北海道    空知郡      中富良野町
54    14621   北海道    空知郡      南富良野町
55    14648   北海道    上川郡        和寒町
56    14656   北海道    上川郡        剣淵町
57    14681   北海道    上川郡        下川町
58    14699   北海道    中川郡        美深町
59    14711   北海道    中川郡        中川町
60    14729   北海道    雨竜郡       幌加内町
61    14818   北海道    増毛郡        増毛町
62    14826   北海道    留萌郡        小平町
63    14834   北海道    苫前郡        苫前町
64    14842   北海道    苫前郡        羽幌町
65    14869   北海道    天塩郡        遠別町
66    14877   北海道    天塩郡        天塩町
67    15121   北海道    枝幸郡       浜頓別町
68    15130   北海道    枝幸郡       中頓別町
69    15148   北海道    枝幸郡        枝幸町
70    15164   北海道    天塩郡        豊富町
71    15172   北海道    礼文郡        礼文町
72    15181   北海道    利尻郡        利尻町
73    15199   北海道    利尻郡      利尻富士町
74    15202   北海道    天塩郡        幌延町
75    15431   北海道    網走郡        美幌町
76    15440   北海道    網走郡        津別町
77    15458   北海道    斜里郡        斜里町
78    15466   北海道    斜里郡        清里町
79    15474   北海道    斜里郡       小清水町
80    15491   北海道    常呂郡       訓子府町
81    15504   北海道    常呂郡        置戸町
82    15521   北海道    常呂郡       佐呂間町
83    15555   北海道    紋別郡        遠軽町
84    15598   北海道    紋別郡        湧別町
85    15601   北海道    紋別郡        滝上町
86    15610   北海道    紋別郡        興部町
87    15636   北海道    紋別郡        雄武町
88    15644   北海道    網走郡        大空町
89    15717   北海道    虻田郡        豊浦町
90    15750   北海道    有珠郡        壮瞥町
91    15784   北海道    白老郡        白老町
92    15814   北海道    勇払郡        厚真町
93    15849   北海道    虻田郡       洞爺湖町
94    15857   北海道    勇払郡        安平町
95    15865   北海道    勇払郡       むかわ町
96    16012   北海道    沙流郡        日高町
97    16021   北海道    沙流郡        平取町
98    16047   北海道    新冠郡        新冠町
99    16071   北海道    浦河郡        浦河町
100   16080   北海道    様似郡        様似町
101   16098   北海道    幌泉郡       えりも町
102   16101   北海道    日高郡      新ひだか町
103   16314   北海道    河東郡        音更町
104   16322   北海道    河東郡        士幌町
105   16331   北海道    河東郡       上士幌町
106   16349   北海道    河東郡        鹿追町
107   16357   北海道    上川郡        新得町
108   16365   北海道    上川郡        清水町
109   16373   北海道    河西郡        芽室町
110   1641

In [20]:
# Wikidata にあって総務省のデータにはないものを表示
# 何もないのが正しい。
df_towns_only_in_wikidata = df_towns_wikidata[~df_towns_wikidata['団体コード'].isin(df_towns['団体コード'])]
df_towns_only_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 郡名, 町名]
Index: []

In [21]:
# 総務省のデータにあって Wikidata にはないものを表示
# 何もないのが正しい。
df_towns_not_in_wikidata = df_towns[~df_towns['団体コード'].isin(df_towns_wikidata['団体コード'])]
df_towns_not_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 市区町村名, 都道府県名
（カナ）, 市区町村名
（カナ）]
Index: []

In [22]:
# 村の一覧を Wikidata から取得
villages_query = """
SELECT DISTINCT ?pref ?prefLabel ?district ?districtLabel ?village ?villageLabel ?code WHERE {
  ?village p:P31 ?statement2 .
  ?statement2 ps:P31 wd:Q4174776 . # ?village は「日本の村」である。
  MINUS { ?statement2 pq:P582 ?endTime. } # 「終了日」の修飾子がないこと。
  MINUS { ?village wdt:P576 ?abolished . } # ただし「廃止日」がないこと。

  OPTIONAL { ?village wdt:P429 ?code . } # 全国地方公共団体コード
  OPTIONAL {
    # 以下のような例で正しく都道府県を得るために、wdt:P131 のあとに '+' が必要。
    # 札幌市 P131 石狩振興局 P131 北海道
    ?village wdt:P131+ ?pref. # ?pref に（直接または間接的に）位置する。
    ?pref wdt:P31 wd:Q50337. # ?pref は「都道府県」である。
    MINUS { ?pref wdt:P576 ?abolished . } # ただし「廃止日」がないこと。
  }
  OPTIONAL {
    ?village p:P131 ?statement1.
    ?statement1 ps:P131 ?district. # ?village は ?district に位置する。
    ?district wdt:P31 wd:Q1122846. # ?district は「郡」である。
    MINUS { ?statement1 pq:P582 ?endTime. } # 「終了日」が修飾子にあるものは除く。
  }

  SERVICE wikibase:label { bd:serviceParam wikibase:language "ja" . }
}
ORDER BY ASC(?code)
"""
json_villages_wikidata = QuerySparql(villages_query)

# pandas DataFrame へ変換
villages_wikidata = []
for item in json_villages_wikidata:
    village = {
        '団体コード': (int(item['code']['value']) if ('code' in item) else -1),
        '都道府県名': (item['prefLabel']['value'] if ('prefLabel' in item) else None),
        '郡名': (item['districtLabel']['value'] if ('districtLabel' in item) else None),
        '町名': (item['villageLabel']['value'] if ('villageLabel' in item) else None)
    }
    villages_wikidata.append(village)
df_villages_wikidata = pd.DataFrame(villages_wikidata)
df_villages_wikidata

団体コード 都道府県名    郡名     町名
0     13048   北海道   石狩郡   新篠津村
1     13919   北海道   島牧郡    島牧村
2     13960   北海道   虻田郡    真狩村
3     13978   北海道   虻田郡   留寿都村
4     14036   北海道   古宇郡     泊村
5     14044   北海道   古宇郡   神恵内村
6     14095   北海道   余市郡   赤井川村
7     14630   北海道   勇払郡    占冠村
8     14702   北海道   中川郡  音威子府村
9     14851   北海道   苫前郡   初山別村
10    15113   北海道   宗谷郡    猿払村
11    15628   北海道   紋別郡   西興部村
12    16381   北海道   河西郡   中札内村
13    16390   北海道   河西郡    更別村
14    16675   北海道   阿寒郡    鶴居村
15    16951   北海道   色丹郡    色丹村
16    16969   北海道   国後郡     泊村
17    16977   北海道   国後郡   留夜別村
18    16985   北海道   択捉郡    留別村
19    16993   北海道   紗那郡    紗那村
20    17001   北海道   蘂取郡    蘂取村
21    23043   青森県  東津軽郡    蓬田村
22    23434   青森県  中津軽郡   西目屋村
23    23671   青森県  南津軽郡   田舎館村
24    24112   青森県   上北郡   六ヶ所村
25    24244   青森県   下北郡    東通村
26    24252   青森県   下北郡   風間浦村
27    24261   青森県   下北郡    佐井村
28    24503   青森県   三戸郡    新郷村
29    34843   岩手県  下閉伊郡   田野畑村
30    34851   岩手県  下閉伊郡    普代村
31    35033   岩手県   九戸郡    野田村
32    35068   岩手県   九戸郡    九戸村
33    44245   宮城県   黒川郡    大衡村
34    53279   秋田県  北秋田郡  上小阿仁村
35    53686   秋田県  南秋田郡    大潟村
36    54640   秋田県   雄勝郡   東成瀬村
37    63657   山形県   最上郡    大蔵村
38    63665   山形県   最上郡    鮭川村
39    63673   山形県   最上郡    戸沢村
40    73229   福島県   安達郡    大玉村
41    73440   福島県   岩瀬郡    天栄村
42    73644   福島県  南会津郡   檜枝岐村
43    74021   福島県   耶麻郡   北塩原村
44    74225   福島県   河沼郡    湯川村
45    74462   福島県   大沼郡    昭和村
46    74616   福島県  西白河郡    西郷村
47    74641   福島県  西白河郡    泉崎村
48    74659   福島県  西白河郡    中島村
49    74845   福島県  東白川郡    鮫川村
50    75027   福島県   石川郡    玉川村
51    75035   福島県   石川郡    平田村
52    75442   福島県   双葉郡    川内村
53    75485   福島県   双葉郡    葛尾村
54    75647   福島県   相馬郡    飯舘村
55    83411   茨城県   那珂郡    東海村
56    84425   茨城県   稲敷郡    美浦村
57   103446   群馬県  北群馬郡    榛東村
58   103667   群馬県   多野郡    上野村
59   103837   群馬県   甘楽郡    南牧村
60   104256   群馬県   吾妻郡    嬬恋村
61   104281   群馬県   吾妻郡    高山村
62   104434   群馬県   利根郡    片品村
63   104442   群馬県   利根郡    川場村
64   104485   群馬県   利根郡    昭和村
65   113697   埼玉県   秩父郡   東秩父村
66   124231   千葉県   長生郡    長生村
67   133078   東京都  西多摩郡    檜原村
68   133621   東京都  None    利島村
69   133639   東京都  None    新島村
70   133647   東京都  None   神津島村
71   133817   東京都  None    三宅村
72   133825   東京都  None   御蔵島村
73   134023   東京都  None   青ヶ島村
74   134210   東京都  None   小笠原村
75   144029  神奈川県   愛甲郡    清川村
76   153427   新潟県  西蒲原郡    弥彦村
77   155047   新潟県   刈羽郡    刈羽村
78   155811   新潟県   岩船郡    関川村
79   155861   新潟県   岩船郡   粟島浦村
80   163210   富山県  中新川郡    舟橋村
81   194221   山梨県  南都留郡    道志村
82   194247   山梨県  南都留郡    忍野村
83   194255   山梨県  南都留郡   山中湖村
84   194298   山梨県  南都留郡    鳴沢村
85   194425   山梨県  北都留郡    小菅村
86   194433   山梨県  北都留郡   丹波山村
87   203041   長野県  南佐久郡    川上村
88   203050   長野県  南佐久郡    南牧村
89   203068   長野県  南佐久郡   南相木村
90   203076   長野県  南佐久郡   北相木村
91   203491   長野県   小県郡    青木村
92   203637   長野県   諏訪郡     原村
93   203858   長野県  上伊那郡   南箕輪村
94   203866   長野県  上伊那郡    中川村
95   203882   長野県  上伊那郡    宮田村
96   204072   長野県  下伊那郡    阿智村
97   204099   長野県  下伊那郡    平谷村
98   204102   長野県  下伊那郡    根羽村
99   204111   長野県  下伊那郡    下條村
100  204129   長野県  下伊那郡    売木村
101  204137   長野県  下伊那郡    天龍村
102  204145   長野県  下伊那郡    泰阜村
103  204153   長野県  下伊那郡    喬木村
104  204161   長野県  下伊那郡    豊丘村
105  204170   長野県  下伊那郡    大鹿村
106  204251   長野県   木曽郡    木祖村
107  204293   長野県   木曽郡    王滝村
108  204307   長野県   木曽郡    大桑村
109  204463   長野県  東筑摩郡    麻績村
110  204480   長野県  東筑摩郡    生坂村
111  204501   長野県  東筑摩郡    山形村
112  204510   長野県  東筑摩郡    朝日村
113  204528   長野県  東筑摩郡    筑北村
114  204820   長野県  北安曇郡    松川村
115  204854   長野県  北安曇郡    白馬村
116  204862   長野県  北安曇郡    小谷村
117  205435   長野県  上高井郡    高山村
118  205621   長野県  下高井郡   木島平村
119  205630   長野県  下高井郡  野沢温泉村
120  205885   長野県  上水内郡    小川村
121  206024   長野県  下水内郡     栄村
122  215074   岐阜県   加茂郡   東白川村
123  216046   岐阜県   大野郡    白川村
124  234273   愛知県   海部郡    飛島村
125  235636   愛知県  北設楽郡    豊根村
126  263672   京都府   相楽郡   南山城村
127  273830   大阪府  南河内郡  千早赤阪村
128  29

In [23]:
# Wikidata にあって総務省のデータにはないものを表示
# 何もないのが正しい。
df_villages_only_in_wikidata = df_villages_wikidata[~df_villages_wikidata['団体コード'].isin(df_villages['団体コード'])]
df_villages_only_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 郡名, 町名]
Index: []

In [24]:
# 総務省のデータにあって Wikidata にはないものを表示
# 何もないのが正しい。
df_Villages_not_in_wikidata = df_villages[~df_villages['団体コード'].isin(df_villages_wikidata['団体コード'])]
df_Villages_not_in_wikidata

Empty DataFrame
Columns: [団体コード, 都道府県名, 市区町村名, 都道府県名
（カナ）, 市区町村名
（カナ）]
Index: []